src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1122,15 +1198,45 @@ private void PerformSyncOperation<TOperation>(ref OperationQueue<TOperation> que
                     return;
                 }
 
-                if (e.Wait(timeout))
+                bool timeoutExpired = false;
+                while (true)
                 {
-                    // Completed within timeout
-                    return;
+                    DateTime waitStart = DateTime.UtcNow;
+
+                    if (!e.Wait(timeout))
+                    {
+                        timeoutExpired = true;
+                        break;
+                    }
+
+                    // Reset the event now to avoid lost notifications if the processing is unsuccessful.
+                    e.Reset();
+
+                    // We've been signalled to try to process the operation.
+                    OperationQueue<TOperation>.OperationResult result = queue.ProcessQueuedOperation(operation);
+                    if (result == OperationQueue<TOperation>.OperationResult.Completed ||
+                        result == OperationQueue<TOperation>.OperationResult.Cancelled)
+                    {
+                        break;
+                    }
+
+                    // Couldn't process the operation.
```

Comment:
In what situations can it happen that Wait returns true (meaning the event was set) but result != Completed && result !=  Canceled such that we end up here?

Reply:
It means we got EAGAIN when trying the operation.  This can happen if, for example, there are two operations in the queue.  First one completes, so we try the second.  It returns EAGAIN.

It may also happen if we get a spurious notification.  In theory this shouldn't happen, but we handle it anyway.